collections/src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,519 @@
+    val elem = this(idx)
+    remove(idx, 1)
+    elem
+  }
+
+  override def subtract(elem: A) = {
+    val idx = indexOf(elem)
+    if (idx >= 0) remove(idx, 1) //TODO: SeqOps should be fluent API
+    this
+  }
+
+  /**
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @return
+    */
+  def removeHeadOption(resizeInternalRepr: Boolean = false): Option[A] =
+    if (isEmpty) None else Some(removeHeadAssumingNonEmpty(resizeInternalRepr))
+
+  /**
+    * Unsafely remove the first element (throws exception when empty)
+    * See also removeHeadOption()
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @throws NoSuchElementException when empty
+    * @return
+    */
+  def removeHead(resizeInternalRepr: Boolean = false): A =
+    if (isEmpty) throw new NoSuchElementException(s"empty collection") else removeHeadAssumingNonEmpty(resizeInternalRepr)
+
+  @inline private[this] def removeHeadAssumingNonEmpty(resizeInternalRepr: Boolean = false): A = {
+    val elem = array(start)
+    array(start) = null
+    start = start_+(1)
+    if (resizeInternalRepr) resize(length)
+    elem.asInstanceOf[A]
+  }
+
+  /**
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @return
+    */
+  def removeLastOption(resizeInternalRepr: Boolean = false): Option[A] =
+    if (isEmpty) None else Some(removeLastAssumingNonEmpty(resizeInternalRepr))
+
+  /**
+    * Unsafely remove the last element (throws exception when empty)
+    * See also removeLastOption()
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @throws NoSuchElementException when empty
+    * @return
+    */
+  def removeLast(resizeInternalRepr: Boolean = false): A =
+    if (isEmpty) throw new NoSuchElementException(s"empty collection") else removeLastAssumingNonEmpty(resizeInternalRepr)
+
+  @inline private[this] def removeLastAssumingNonEmpty(resizeInternalRepr: Boolean = false): A = {
+    end = end_-(1)
+    val elem = array(end)
+    array(end) = null
+    if (resizeInternalRepr) resize(length)
+    elem.asInstanceOf[A]
+  }
+
+  /**
+    * Remove all elements from this collection and return the elements while emptying this data structure
+    * @return
+    */
+  def removeAll(): strawman.collection.Seq[A] = {
+    val elems = strawman.collection.Seq.newBuilder[A]
+    elems.sizeHint(length)
+    while(nonEmpty) {
+      elems += removeHeadAssumingNonEmpty()
+    }
+    elems.result()
+  }
+
+  /**
+    * Returns and removes all elements from the left of this queue which satisfy the given predicate
+    *
+    *  @param f   the predicate used for choosing elements
+    *  @return
+    */
+  def removeHeadWhile(f: A => Boolean): strawman.collection.Seq[A] = {
```

Comment:
Since we have `take` and `takeRight`, `drop` and `dropRight`, maybe we should follow the same convention here and have `removeWhile` and `removeRightWhile`?